File: app/src/main/java/com/example/appgreenflow/ui/home/HomeViewModel.java
package com.example.appgreenflow.ui.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.appgreenflow.Article;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;

import java.util.ArrayList;
import java.util.List;

public class HomeViewModel extends ViewModel {
    private final MutableLiveData<List<Article>> articles = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private FirebaseFirestore db;
    private DocumentSnapshot lastVisible;  // For pagination
    private boolean isFirstLoad = true;

    public HomeViewModel() {
        db = FirebaseFirestore.getInstance();
    }

    public LiveData<List<Article>> getArticles() {
        return articles;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void loadArticles(boolean loadMore) {
        if (isLoading.getValue() == true) return;  // Prevent duplicate loads

        isLoading.setValue(true);
        errorMessage.setValue(null);

        Query query = db.collection("articles").orderBy("timestamp", Query.Direction.DESCENDING).limit(10);
        if (!isFirstLoad && lastVisible != null && loadMore) {
            query = query.startAfter(lastVisible);  // Pagination
        }

        query.get()
                .addOnSuccessListener(querySnapshot -> {
                    List<Article> currentList = articles.getValue();
                    if (currentList == null) currentList = new ArrayList<>();

                    for (DocumentSnapshot doc : querySnapshot.getDocuments()) {
                        Article article = doc.toObject(Article.class);
                        if (article != null) {
                            currentList.add(article);
                        }
                    }

                    articles.setValue(currentList);
                    if (!querySnapshot.isEmpty()) {
                        lastVisible = querySnapshot.getDocuments().get(querySnapshot.size() - 1);  // Next startAfter
                    }

                    isFirstLoad = false;
                    isLoading.setValue(false);
                })
                .addOnFailureListener(e -> {
                    errorMessage.setValue("Lỗi load bài báo: " + e.getMessage());
                    isLoading.setValue(false);
                });
    }

    // Reset for refresh
    public void refreshArticles() {
        articles.setValue(new ArrayList<>());
        lastVisible = null;
        isFirstLoad = true;
        loadArticles(false);
    }
}

File: app/src/main/java/com/example/appgreenflow/ui/home/HomeViewModel.kt
package com.example.appgreenflow.ui.home

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore

class HomeViewModel : ViewModel() {
    private val articles: MutableLiveData<MutableList<Article?>?> =
        MutableLiveData(ArrayList<Article?>())
    private val isLoading: MutableLiveData<Boolean?> =
        MutableLiveData(false)
    private val errorMessage: MutableLiveData<String?> = MutableLiveData()
    private val db: FirebaseFirestore
    private var lastVisible: DocumentSnapshot? = null
    private var isFirstLoad = true

    init {
        db = FirebaseFirestore.getInstance()
    }

    fun getArticles(): LiveData<MutableList<Article?>?> {
        return articles
    }

    fun getIsLoading(): LiveData<Boolean?> {
        return isLoading
    }

    fun getErrorMessage(): LiveData<String?> {
        return errorMessage
    }

    fun loadArticles(loadMore: Boolean) {
        if (isLoading.getValue() == true) return

        isLoading.value = true
        errorMessage.value = null

        var query = db.collection("articles")
            .orderBy("timestamp", com.google.firebase.firestore.Query.Direction.DESCENDING)
            .limit(10)
        if (!isFirstLoad && lastVisible != null && loadMore) {
            query = query.startAfter(lastVisible)
        }

        query.get()
            .addOnSuccessListener { querySnapshot ->
                var currentList: MutableList<Article?>? = articles.value
                if (currentList == null) currentList = ArrayList()

                for (doc in querySnapshot.documents) {
                    val article = doc.toObject(Article::class.java)
                    if (article != null) {
                        currentList.add(article)
                    }
                }

                articles.value = currentList
                if (!querySnapshot.isEmpty) {
                    lastVisible = querySnapshot.documents[querySnapshot.size() - 1]
                }

                isFirstLoad = false
                isLoading.value = false
            }
            .addOnFailureListener { e ->
                errorMessage.value = "Lỗi load bài báo: ${e.message}"
                isLoading.value = false
            }
    }

    fun refreshArticles() {
        articles.value = ArrayList()
        lastVisible = null
        isFirstLoad = true
        loadArticles(false)
    }
}